### PR TITLE
DESK-899 NSIS installer update for 2.9.4 to 2.9.5 update bug

### PR DESCRIPTION
**Pull request by JamieRuderman**
### Description

NSIS script update to support keeping config file during update from 2.9.4 to 2.9.5 on windows. 
- Also supporting update of older 2.9.x build and any other older build.

### Ticket

<https://remoteit.atlassian.net/browse/DESK-899 >

### CODE REVIEW

File: installer.nsh [hunk 1]
```diff
@@ -18,9 +18,16 @@
     ; stop the agent
     nsExec::ExecToStack /OEM 'powershell "& " "$\'"$path_i\remoteit.exe$\'" -j agent stop'
 
-    ; move the config file and connections to backup location (protect against 2.9.2 uninstall bug)
-    Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json"
-    Rename "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\connections"
+    ; move the config file and connections to backup location ONLY MOVES IF EMPTY (protect against 2.9.2 uninstall bug)
```
Comment (JamieRuderman) [inline installer.nsh:21]: Delete any existing configs found
Then do the move instead of copying

File: installer.nsh [hunk 2]
```diff
@@ -101,24 +111,25 @@
 
     FileWrite $uninstallLog "$\n***** Remove Files *****$\r$\n"
 
-    ; stop the agent
-    nsExec::ExecToStack /OEM 'powershell "& " "$\'"$path_u\remoteit.exe$\'" -j agent stop'
-
```
Comment (JamieRuderman) [inline installer.nsh:106]: Move stoping service to after copy

File: installer.nsh [hunk 3]
```diff
@@ -18,9 +18,16 @@
     ; stop the agent
     nsExec::ExecToStack /OEM 'powershell "& " "$\'"$path_i\remoteit.exe$\'" -j agent stop'
 
-    ; move the config file and connections to backup location (protect against 2.9.2 uninstall bug)
-    Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json"
```
Comment (penngrove) [inline installer.nsh:22]: First we delete ${REMOTEIT_BACKUP}\config.json.2.9.5 if it exists
then we can safely move
Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json.2.9.5"

File: installer.nsh [hunk 4]
```diff
@@ -18,9 +18,16 @@
...
-    Rename "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\connections"
```
Comment (penngrove) [inline installer.nsh:23]: same for connections

File: installer.nsh [hunk 5]
```diff
@@ -18,9 +18,16 @@
...
+    ; move the config file and connections to backup location ONLY MOVES IF EMPTY (protect against 2.9.2 uninstall bug)
+    ; Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json"
+    ; Rename "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\connections"
+
+    ; copy the config file and connections to backup location
+    CopyFiles /SILENT "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\"
```
Comment (penngrove) [inline installer.nsh:26]: You have to get the config.json out of /etc/remoteit OR 2.9.0 installer will unregister the devices

Comment (penngrove) [reply installer.nsh:26]: No use move and NOT copy

File: installer.nsh [hunk 6]
```diff
@@ -101,24 +111,25 @@
...
     ; create backup directory
     CreateDirectory "${REMOTEIT_BACKUP}"
 
-    ; copy the config file to backup location
+    ; MessageBox MB_OK "uninstall copy files" 
+    ; copy the config file to backup location - possible enhancements:
+    ; - only copy on user request
+    ; - check for a non empty file before copy
+    ; - remove altogether
     CopyFiles /SILENT "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\"
```
Comment (penngrove) [inline installer.nsh:122]: Use CopyFiles /SILENT "$APPDATA\remoteit\config.json.2.9.5" "${REMOTEIT_BACKUP}\"

File: installer.nsh [hunk 7]
```diff
@@ -56,15 +63,18 @@
...
     StrCpy $ps_command 'powershell "& " "$\'"$path_\remoteit.exe$\'" -j agent uninstall'
     nsExec::ExecToStack /OEM $ps_command
     Pop $0
     Pop $1
     FileWrite $installLog "$ps_command     [$0]  $1$\r$\n"
 
-    ; restore config from backup
+    ; restore config from backup - question: should this always happen? Could be error prone.
+    ; MessageBox MB_OK "Install: will restore files"
     CopyFiles /SILENT "${REMOTEIT_BACKUP}\config.json" "$APPDATA\remoteit\"
```
Comment (penngrove) [inline installer.nsh:75]: CopyFiles /SILENT "${REMOTEIT_BACKUP}\config.json.2.9.5" "$APPDATA\remoteit\"

File: installer.nsh [hunk 8]
```diff
@@ -101,24 +111,25 @@
 
     FileWrite $uninstallLog "$\n***** Remove Files *****$\r$\n"
 
-    ; stop the agent
-    nsExec::ExecToStack /OEM 'powershell "& " "$\'"$path_u\remoteit.exe$\'" -j agent stop'
```
Comment (penngrove) [inline installer.nsh:105]: Move the CLI stop to AFTER we backup the config files

File: installer.nsh (hunk 6 above)
Comment (penngrove) [inline installer.nsh:122]: Remove these lines, no need to backup likely empty files

File: installer.nsh (hunk 3 above)
Comment (penngrove) [reply installer.nsh:22]: Would be nice to automate this file name interpolation with a $version or something

Comment (penngrove) [reply installer.nsh:22]: if the file exists it is likely a good backup and we should not delete it or replace with anything

File: installer.nsh [hunk 9]
```diff
@@ -18,9 +18,16 @@
...
-    ; move the config file and connections to backup location (protect against 2.9.2 uninstall bug)
-    Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json"
-    Rename "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\connections"
+    ; move the config file and connections to backup location ONLY MOVES IF EMPTY (protect against 2.9.2 uninstall bug)
+    ; Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json"
+    ; Rename "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\connections"
+
+    ; copy the config file and connections to backup location
+    CopyFiles /SILENT "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\"
+    CopyFiles /SILENT "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\"
```
Comment (penngrove) [inline installer.nsh:27]: Remove the copy, we should move

File: installer.nsh (hunk 7 above)
Comment (penngrove) [reply installer.nsh:75]: Always restore hte correct versioned backup

File: installer.nsh (hunk 8 above)
Comment (penngrove) [inline installer.nsh:105]: We do not need to stop the agent

File: installer.nsh [hunk 10]
```diff
@@ -101,24 +111,25 @@
...
     CreateDirectory "${REMOTEIT_BACKUP}"
 
-    ; copy the config file to backup location
+    ; MessageBox MB_OK "uninstall copy files" 
+    ; copy the config file to backup location - possible enhancements:
+    ; - only copy on user request
+    ; - check for a non empty file before copy
+    ; - remove altogether
     CopyFiles /SILENT "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\"
     CopyFiles /SILENT "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\"
```
Comment (penngrove) [inline installer.nsh:123]: These two copy are of little or no value and can be removed

File: installer.nsh (hunk 3 above)
Comment (JamieRuderman) [reply installer.nsh:22]: Maybe we should delete it first... so if they re-install again after setting stuff up / changing thing it won't restore an old file. Or we should move it back and not leave it in backup... make it just a temp file.

File: installer.nsh [hunk 11]
```diff
@@ -18,9 +19,12 @@
     ; stop the agent
     nsExec::ExecToStack /OEM 'powershell "& " "$\'"$path_i\remoteit.exe$\'" -j agent stop'
 
-    ; move the config file and connections to backup location (protect against 2.9.2 uninstall bug)
-    Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json"
-    Rename "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\connections"
+    ; move the config file and connections to backup location ONLY MOVES IF EMPTY (protect against 2.9.2 uninstall bug)
+    Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config${PKGVERSION}.json"
```
Comment (penngrove) [inline installer.nsh:23]: instead of config${PKGVERSION} I would prefer config-${PKGVERSION}

File: installer.nsh [hunk 12]
```diff
@@ -56,15 +60,18 @@
...
     Pop $0
     Pop $1
     FileWrite $installLog "$ps_command     [$0]  $1$\r$\n"
 
-    ; restore config from backup
-    CopyFiles /SILENT "${REMOTEIT_BACKUP}\config.json" "$APPDATA\remoteit\"
-    CopyFiles /SILENT "${REMOTEIT_BACKUP}\connections" "$PROFILE\AppData\Local\remoteit\"
+    ; restore config from backup - question: should this always happen? Could be error prone.
+    ; MessageBox MB_OK "Install: will restore files"
+    CopyFiles /SILENT "${REMOTEIT_BACKUP}\config${PKGVERSION}.json" "$APPDATA\remoteit\config.json"
```
Comment (penngrove) [inline installer.nsh:72]: should use a dash config-${PKGVERSION}.json

Comment (penngrove) [reply installer.nsh:72]: Try a move here

File: installer.nsh (hunk 11 above)
Comment (penngrove) [reply installer.nsh:23]: delete destination 2.9.5 (if it exists) and move (better)

File: installer.nsh [hunk 13]
```diff
@@ -12,15 +13,21 @@
...
 
-    ; move the config file and connections to backup location (protect against 2.9.2 uninstall bug)
-    Rename "$APPDATA\remoteit\config.json" "${REMOTEIT_BACKUP}\config.json"
-    Rename "$PROFILE\AppData\Local\remoteit\connections" "${REMOTEIT_BACKUP}\connections"
+    ; create backup directory if doesn't exist
+    CreateDirectory "${REMOTEIT_BACKUP}"
+
+    ; remove old backups so the move can occur
+    Delete "${REMOTEIT_BACKUP}\config-${PKGVERSION}.json"
+    RMDir /r  "${REMOTEIT_BACKUP}\connections-${PKGVERSION}"
```
Comment (penngrove) [inline installer.nsh:24]: I don't get the rmdir. It's not a directory, is it

**Review by penngrove — APPROVED** [review]

**Pull request merged**